themes/dotmatrix/dotmatrix.theme

```diff
@@ -0,0 +1,1053 @@
+                __dotmatrix_make_sep_mid ${DOTMATRIX_COLOR_SEP_FAIR[@]+"${DOTMATRIX_COLOR_SEP_FAIR[@]}"}
+                _DOTMATRIX_VCS+=" $dotmatrix_sep "
+            else
+                __dotmatrix_make_sep_plug ${DOTMATRIX_COLOR_LITE[@]+"${DOTMATRIX_COLOR_LITE[@]}"}
+                _DOTMATRIX_VCS+="$dotmatrix_sep"
+            fi
+        fi
+
+        if [[ -n "$stash" ]] ; then
+            __dotmatrix_make_field "$stash" ${DOTMATRIX_COLOR_LITE[@]+"${DOTMATRIX_COLOR_LITE[@]}"}
+            _DOTMATRIX_VCS+="$dotmatrix_field"
+            if [[ -n "$untracked" ]] ; then
+                __dotmatrix_make_field "$untracked" ${DOTMATRIX_COLOR_LITE[@]+"${DOTMATRIX_COLOR_LITE[@]}"}
+                __dotmatrix_make_sep_soft ${DOTMATRIX_COLOR_SEP_LITE[@]+"${DOTMATRIX_COLOR_SEP_LITE[@]}"}
+                _DOTMATRIX_VCS+="${dotmatrix_sep}${dotmatrix_field}"
+            fi
+            if [[ -n "$head" ]] ; then
+                __dotmatrix_make_field "$head" ${DOTMATRIX_COLOR_LITE[@]+"${DOTMATRIX_COLOR_LITE[@]}"}
+                __dotmatrix_make_sep_soft ${DOTMATRIX_COLOR_SEP_LITE[@]+"${DOTMATRIX_COLOR_SEP_LITE[@]}"}
+                _DOTMATRIX_VCS+="${dotmatrix_sep}${dotmatrix_field}"
+            fi
+        else # no stash
+            if [[ -n "$untracked" ]] ; then
+                __dotmatrix_make_field "$untracked" ${DOTMATRIX_COLOR_LITE[@]+"${DOTMATRIX_COLOR_LITE[@]}"}
+                _DOTMATRIX_VCS+="${dotmatrix_field}"
+                if [[ -n "$head" ]] ; then
+                    __dotmatrix_make_field "$head" ${DOTMATRIX_COLOR_LITE[@]+"${DOTMATRIX_COLOR_LITE[@]}"}
+                    __dotmatrix_make_sep_soft ${DOTMATRIX_COLOR_SEP_LITE[@]+"${DOTMATRIX_COLOR_SEP_LITE[@]}"}
+                    _DOTMATRIX_VCS+="${dotmatrix_sep}${dotmatrix_field}"
+                fi
+            else # no untracked
+                if [[ -n "$head" ]] ; then
+                    __dotmatrix_make_field "$head" ${DOTMATRIX_COLOR_LITE[@]+"${DOTMATRIX_COLOR_LITE[@]}"}
+                    _DOTMATRIX_VCS+="${dotmatrix_field}"
+                fi
+            fi
+        fi # stash
+        return 1
+    else
+        _DOTMATRIX_VCS=""
+        return 0
+    fi # find_vcs
+}
+
+########################################################################
+# Helper functions
+########################################################################
+
+__dotmatrix_join_fields() {
+    dotmatrix_join_fields=""
+    local sep="${1-}"
+    shift
+    if [[ -n "$@" ]] ; then
+        local first_field="${1-}"
+        shift
+        dotmatrix_join_fields="${first_field}"
+        for f in "$@"; do
+            dotmatrix_join_fields+="${sep}${f}"
+        done
+    fi
+}
+
+__dotmatrix_make_field() { # string, (color array)
+    local lp_terminal_format string="${1-}"
+    [[ -z $string ]] && return 1
+
+    shift
+    local format="$@"
+    if [[ -z $format ]] ; then
+        lp_terminal_format -1
+    else
+        lp_terminal_format $format
```

Comment:
This should not work. If it is working as you expect, it is a lucky bug, but I doubt it will work correctly on all shells.

`var="$@"` does _not_ make `var` an array. Much better to just call `lp_terminal_format "$@"`, that will work exactly as expected.

I would also skip the 
```bash
    if [[ -z $format ]] ; then
        lp_terminal_format -1
```
part, this is a private function to the theme, just enforce the method signature of always being explicit with what formatting you want.